### ISSUE TITLE
clang-static-analyzer: Aggregate-initialized struct field incorrectly marked undefined

### ISSUE THREAD

**Issue by rojer** [llvm/llvm-project#116444]
This code:

```cpp
#include <iostream>
#include <string>

struct Structy {
  std::string foo;
  int bar;
};

int main() {
  auto *s = new Structy{
      .foo = "foo",
      .bar = 1,
  };
  std::cout << s->foo << s->bar;
  delete s;
}
```

produces the following static analyzer warning:

```
$ clang-tidy -p . qq.cpp 
1 warning generated.
/home/rojer/allterco/shelly-ng3/qq.cpp:14:3: error: 1st function call argument is an uninitialized value [clang-analyzer-core.CallAndMessage,-warnings-as-errors]
   14 |   std::cout << s->foo << s->bar;
      |   ^                      ~~~~~~
/home/rojer/allterco/shelly-ng3/qq.cpp:10:13: note: Uninitialized value stored to field 'bar'
   10 |   auto *s = new Structy{
      |             ^~~~~~~~~~~~
   11 |       .foo = "foo",
      |       ~~~~~~~~~~~~~
   12 |       .bar = 1,
      |       ~~~~~~~~~
   13 |   };
      |   ~
/home/rojer/allterco/shelly-ng3/qq.cpp:14:3: note: 1st function call argument is an uninitialized value
   14 |   std::cout << s->foo << s->bar;
      |   ^                      ~~~~~~
1 warning treated as error
```

`s->bar` is of course explicitly initialized to 1, so the generated warning is a false-positive.

swapping field order to bar, foo makes it go away.

tested with latest main (20.0.0git ec353b7418e272e96cd63cc61bec586ab49da92f).

**Comment by vinay-deshmukh** [comment]
I'm working on this issue.

**Comment by rojer** [comment]
ping @vinay-deshmukh any progress?

**Comment by vinay-deshmukh** [comment]
Hi @rojer , sorry haven't been able to come back to work/make progress on this issue!

**Comment by steakhal** [comment]
> Picking this one up if okay with you [@vinay-deshmukh](https://github.com/vinay-deshmukh).
> 
> I made a slightly more minimzed reproducer:
> 
> ```
> struct S {
>   int foo;
>   int bar;
> };
> 
> void f() {
>   S *s = new S{
>       .foo = 13,
>       .bar = 1,
>   };
>   int retval = s->bar;
>   delete s;
> }
> ```
> 
> Looks like a problem with CXXNewExpr with the init lists. The results even when swapping the field order are still not as good as could be. Going to work on aggregate initialization semantics in exprengine I think...

Have a look at `ExprEngine::ProcessInitializer`.

**Comment by vinay-deshmukh** [comment]
> Picking this one up if okay with you [@vinay-deshmukh](https://github.com/vinay-deshmukh).
> 
> I made a slightly more minimzed reproducer:
> 
> ```
> struct S {
>   int foo;
>   int bar;
> };
> 
> void f() {
>   S *s = new S{
>       .foo = 13,
>       .bar = 1,
>   };
>   int retval = s->bar;
>   delete s;
> }
> ```
> 
> Looks like a problem with CXXNewExpr with the init lists. The results even when swapping the field order are still not as good as could be. Going to work on aggregate initialization semantics in exprengine I think...

Go for it!

**Comment by rojer** [comment]
thanks a lot @Flandini !
